Contributing the Module Mismatch Simulation script

bmeyers opened the pull request.

bmeyers commented:
@mikofski ready for review

bmeyers commented:
@mikofski you still interested in adding this to v3 as well?

mikofski commented:
Yep, but I got other fish to fry. Is there a rush on it? Do you need this?

bmeyers commented:
nope, no rush. just wanted to get it out there. did you want to review the code first?

bmeyers commented:
@mikofski can i merge this, or should there be a review first?

mikofski commented:
IMO, there should always be at least one review. What's your opinion?

bmeyers commented:
Fine by me, I just wasn't sure what you had in mind in terms of a review for a stand-alone contribution like this.

bmeyers commented:
on it

mikofski commented on pvmismatch/contrib/ModuleMismatchSimulator.py:
```diff
@@ -0,0 +1,510 @@
+    import sys
+    sys.path.append('../../')
+    from pvmismatch import PVsystem, PVmodule, PVcell, PVconstants
+    from pvmismatch.pvmismatch_lib.pvmodule import STD72, STD96, STD128
```
Can you drop this, and just import pvmismatch? I don't want to endorse altering `sys.path` in code. This can be done from CLI using `pip install -e path/to/pvmismatch` or `export PYTHONPATH=path/to/pvmismatch:$PYTHONPATH`. Ideally `sys.path` should only be used in an interpreter, not in a script or package module. IMO runtime alterations of path are a ["code smell"][1] that should be avoided.

[1]: https://en.wikipedia.org/wiki/Code_smell

---
https://blog.codinghorror.com/code-smells/
http://wiki.c2.com/?CodeSmell
https://sourcemaking.com/refactoring/smells
https://sourcemaking.com/antipatterns
https://en.wikipedia.org/wiki/Code_smell

bmeyers commented:
Looks clean to me know. Only difference from master is the new script.

mikofski commented on pvmismatch/contrib/ModuleMismatchSimulator.py:
```diff
@@ -0,0 +1,510 @@
+'''
```
Please add a shebang so I don't have to type `python ModuleMismatchSimulatory.py` to start it.

mikofski replied:
Also, can you change the name to follow [pep8 package and module naming conventions](https://www.python.org/dev/peps/pep-0008/#package-and-module-names)? thanks!

bmeyers replied:
some information on shebang best practices in Python scripts: http://stackoverflow.com/questions/6908143/should-i-put-shebang-in-python-scripts-and-what-form-should-it-take

bmeyers replied:
`#!/usr/bin/env python` seems to be the best practice, but I am having problems testing it on my computer. I have the shebang in the script, and it runs fine when I call python first, but I get an error if I just try to call the script:

```
(pvmismatch_dev) contrib [bennetmeyers] $ ModuleMismatchSimulator.py 
-bash: ModuleMismatchSimulator.py: command not found
```

Am I missing something?

mikofski replied:
BaSH is very picky about what is executable. Try (note the **dot + forward slash**):

    $ ./ModuleMismatchSimulator.py

works for me.

Also on a *nix system, the executable bit would also have to be flipped. Since you are on Mac OSX can you do this?

    $ chmod 755 ModuleMismatchSimulator.py

this will make it executable by everyone

type `ls -l ModuleMismatchSimulator.py` to see _all_ the file attributes.

bmeyers replied:
Yup, that's working for me now

mikofski replied:
did you change the executable bit? you'll need to commit and push that for it to stick otherwise every *nix user will have to do it themselves. thanks!

mikofski replied:
It's showing up as executable, `-rwxr-xr-x` (= 755) on my machine, but I'm on windows, so not sure if it is on yours too.
![already_executable](https://cloud.githubusercontent.com/assets/1385621/22844452/37e75be2-ef93-11e6-940d-c28df6e33073.png)

bmeyers replied:
Yeah, that was wrapped in the 'typo fix' commit. Sorry, should have identified that change as well. Working on the rest, but need to take care of something else first.

mikofski commented:
Yep, it's much easier to read now. Thanks, I'm reviewing it now.

mikofski commented:
Unfortunately it doesn't seem to work with the latest matplotlib and numpy?
matplotlib (2.0.0)
numpy (1.12.0+mkl)

Here's the [traceback](http://pastebin.com/jPQR3Eyy)

Here's a screenshot:
![pvmismatch_module_simulator](https://cloud.githubusercontent.com/assets/1385621/22838643/07b7034a-ef7b-11e6-9a52-ea569a21db0b.png)

bmeyers commented:
Interesting, I will double check my version of both and do some testing. 

And i hear you loud and clear on the issue of altering path. I'll simply have the exception inform the user that pvmismatch needs to be on their path before they can run the script and suggest the solutions you describe.

mikofski commented:
IMO, you shouldn't need that `try: ... except: ...` at all because if the import fails, then Python will already raise an `ImportError`. We want an exception here, and it's already being handled, and it actually already tells the user everything they need to know to fix the problem

bmeyers commented:
I just liked the idea of giving a little feedback to the user. In my early days, I had some confusion on that point, and I thought what your wrote was helpful.

mikofski commented on pvmismatch/contrib/ModuleMismatchSimulator.py:
```diff
@@ -0,0 +1,512 @@
+    b_saveshade.on_clicked(save_the_shade)
+    b_clearlast.on_clicked(ClearLast)
+
+plt.show()
```
please add a newline character to the end too, thanks

mikofski replied:
also please change the name of the module to conform to pep8:
>Modules should have short, all-lowercase names. Underscores can be used in the module name if it improves readability. Python packages should also have short, all-lowercase names, although the use of underscores is discouraged.

Something all lower case like `simulator.py` or `mismatch-simulator.py` - since it's a script it's okay to have a dash in it, just means you can't import it from another module, which you'll never need to do anyway, unless you want to in which case, do **not** put a dash in the name, use underscore instead like `module_mismatch_simulator.py`. Whatever you think is best, just not camel case, save that for classes.

bmeyers replied:
done

mikofski commented on pvmismatch/contrib/ModuleMismatchSimulator.py:
```diff
@@ -33,7 +33,7 @@
-    print "or export 'PYTHONPATH=path/to/pvmismatch:$PYTHONPATH' first."
+    print "or 'export PYTHONPATH=path/to/pvmismatch:$PYTHONPATH' first."
     import sys
     sys.exit()
```
since this is an error, change the return to -1

    sys.exit(-1)

bmeyers replied:
done

bmeyers commented:
@mikofski I believe I have addressed all the comments. Thank you for giving this such a thorough review! I appreciate the feedback. Learning a lot over here.

mikofski commented:
OK @bmeyers can you tag this as v3.0 using git and push the tags to upstream

    $ git push --tags upstream

then can you build a new wheel and a new tarball and upload them to pypi.

    (venv) $ python setup.py bdist_wheel sdist upload

there is no need to update the docs at all, so leave them alone.

you might want to upload the wheel and tarball to github releases as well, they're in `dist/`

then let's work on issue #49
